core/funcs.go

```diff
@@ -852,6 +896,150 @@ func init() {
+			ArgTypes:         [][]RslTypeEnum{{RslFloatT, RslIntT}},
+			NamedArgs:        NO_NAMED_ARGS,
+			Execute: func(f FuncInvocationArgs) []RslValue {
+				arg := f.args[0]
+				val := arg.value.RequireFloatAllowingInt(f.i, arg.node)
+				return newRslValues(f.i, f.callNode, math.Ceil(val))
+			},
+		},
+		{
+			Name:             FUNC_FLOOR,
+			ReturnValues:     ONE_RETURN_VAL,
+			RequiredArgCount: 1,
+			ArgTypes:         [][]RslTypeEnum{{RslFloatT, RslIntT}},
+			NamedArgs:        NO_NAMED_ARGS,
+			Execute: func(f FuncInvocationArgs) []RslValue {
+				arg := f.args[0]
+				val := arg.value.RequireFloatAllowingInt(f.i, arg.node)
+				return newRslValues(f.i, f.callNode, math.Floor(val))
+			},
+		},
+		{
+			Name:             FUNC_MIN,
+			ReturnValues:     ONE_RETURN_VAL,
+			RequiredArgCount: 1,
+			ArgTypes:         [][]RslTypeEnum{{RslListT}},
+			NamedArgs:        NO_NAMED_ARGS,
+			Execute: func(f FuncInvocationArgs) []RslValue {
+				// input is a list of numbers
+				list := f.args[0].value.RequireList(f.i, f.args[0].node)
+				if len(list.Values) == 0 {
+					f.i.errorf(f.callNode, "Cannot find minimum of empty list")
+				}
+
+				minVal := math.MaxFloat64
+				for idx, item := range list.Values {
+					val, ok := item.TryGetFloatAllowingInt()
+					if !ok {
+						f.i.errorf(f.args[0].node, "%s() requires a list of numbers, got %q at index %d", FUNC_MIN, TypeAsString(item), idx)
+					}
+					minVal = math.Min(minVal, val)
+				}
+				return newRslValues(f.i, f.callNode, minVal)
+			},
+		},
+		{
+			Name:             FUNC_MAX,
+			ReturnValues:     ONE_RETURN_VAL,
+			RequiredArgCount: 1,
+			ArgTypes:         [][]RslTypeEnum{{RslListT}},
+			NamedArgs:        NO_NAMED_ARGS,
+			Execute: func(f FuncInvocationArgs) []RslValue {
+				// input is a list of numbers
+				list := f.args[0].value.RequireList(f.i, f.args[0].node)
+				if len(list.Values) == 0 {
+					f.i.errorf(f.callNode, "Cannot find maximum of empty list")
+				}
+
+				maxVal := -math.MaxFloat64
+				for idx, item := range list.Values {
+					val, ok := item.TryGetFloatAllowingInt()
+					if !ok {
+						f.i.errorf(f.args[0].node, "%s() requires a list of numbers, got %q at index %d", FUNC_MAX, TypeAsString(item), idx)
+					}
+					maxVal = math.Max(maxVal, val)
+				}
+				return newRslValues(f.i, f.callNode, maxVal)
+			},
+		},
+		{
+			Name:             FUNC_CLAMP,
+			ReturnValues:     ONE_RETURN_VAL,
+			RequiredArgCount: 1,
+			ArgTypes:         [][]RslTypeEnum{{RslFloatT, RslIntT}, {RslFloatT, RslIntT}, {RslFloatT, RslIntT}},
+			NamedArgs:        NO_NAMED_ARGS,
+			Execute: func(f FuncInvocationArgs) []RslValue {
+				// input is a number and a min and max
+				val, ok := f.args[0].value.TryGetFloatAllowingInt()
+				if !ok {
+					f.i.errorf(f.args[0].node, "%s() requires a number, got %q at index %d", FUNC_CLAMP, TypeAsString(f.args[0].value), 0)
+				}
+				min := -math.MaxFloat64
+				max := math.MaxFloat64
+				switch len(f.args) {
```

Comment:
Rather than having this validation, you can instead set `RequiredArgCount` to 3. The shared code before invoking `Execute` will thus guarantee that we get all 3 args we expect, so you can safely start this function with something like

```go
numArg := f.args[0]
minArg := f.args[1]
maxArg := f.args[2]
```

and remove the switch validation below.

Also saves you doing the comment on line 1001 because the above ^ variables make that clear (self-documenting code).